### ISSUE TITLE
Mapped JSON in multiset causes multiple ad-hoc conversions

### ISSUE THREAD

**Issue by shimikano** [jOOQ/jOOQ#12697]
I know that JSON content in an multiset emulated with JSON may not yet officially be fully supported (#12021), but we ran into the following issue with a mapped JSON field inside a multiset:

```java

public static class Foo {

  private final String a;

  public Foo(@JsonProperty("a") String a) { this.a = a; }

  @Override
  public String toString() { return "Foo{a=" + a + "}"; }

}

// helper method, using Jackson
private static <T> T parseJson(JSON json, Class<T> type) {
  try {
    return new ObjectMapper().readValue(json.data(), type);
  } catch (JsonProcessingException e) {
    throw new UncheckedIOException(e);
  }
}

@Test
public void multipleConversionsInMultisetWithMappedJson() {
  create.select(
    multiset(
      select(
        // a JSON field mapped to Foo
        jsonObject(key("a").value("hello")).convertFrom(json -> parseJson(json, Foo.class))
      )
    )
  )
  .fetch();
}
```

### Expected behavior

No error.

### Actual behavior

Fails with:

```
org.jooq.exception.DataAccessException: SQL [select (
  select coalesce(
    jsonb_agg(jsonb_build_array(v0)),
    jsonb_build_array()
  )
  from (
    select json_build_object(?, cast(? as varchar)) as v0
  ) as t
)]; Error while reading field: multiset(
  select json_object(key 'a' value 'hello')
), at JDBC index: 1
	at org.jooq_3.15.4.POSTGRES.debug(Unknown Source) ~[na:na]
	at org.jooq.impl.Tools.translate(Tools.java:2988) ~[jooq-3.15.4.jar:na]
	at org.jooq.impl.DefaultExecuteContext.sqlException(DefaultExecuteContext.java:639) ~[jooq-3.15.4.jar:na]
	at org.jooq.impl.CursorImpl$CursorIterator.fetchNext(CursorImpl.java:1439) ~[jooq-3.15.4.jar:na]
	at org.jooq.impl.CursorImpl$CursorIterator.hasNext(CursorImpl.java:1400) ~[jooq-3.15.4.jar:na]
	at org.jooq.impl.CursorImpl.fetchNext(CursorImpl.java:220) ~[jooq-3.15.4.jar:na]
	at org.jooq.impl.AbstractCursor.fetch(AbstractCursor.java:177) ~[jooq-3.15.4.jar:na]
	at org.jooq.impl.AbstractCursor.fetch(AbstractCursor.java:88) ~[jooq-3.15.4.jar:na]
	at org.jooq.impl.AbstractResultQuery.execute(AbstractResultQuery.java:259) ~[jooq-3.15.4.jar:na]
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:335) ~[jooq-3.15.4.jar:na]
	at org.jooq.impl.AbstractResultQuery.fetch(AbstractResultQuery.java:284) ~[jooq-3.15.4.jar:na]
	at org.jooq.impl.SelectImpl.fetch(SelectImpl.java:2842) ~[jooq-3.15.4.jar:na]
[..]
Caused by: java.sql.SQLException: Error while reading field: multiset(
  select json_object(key 'a' value 'hello')
), at JDBC index: 1
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.setValue(CursorImpl.java:1561) ~[jooq-3.15.4.jar:na]
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.apply(CursorImpl.java:1500) ~[jooq-3.15.4.jar:na]
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.apply(CursorImpl.java:1459) ~[jooq-3.15.4.jar:na]
	at org.jooq.impl.RecordDelegate.operate(RecordDelegate.java:143) ~[jooq-3.15.4.jar:na]
	at org.jooq.impl.CursorImpl$CursorIterator.fetchNext(CursorImpl.java:1424) ~[jooq-3.15.4.jar:na]
	... 92 common frames omitted
Caused by: java.lang.RuntimeException: java.io.UncheckedIOException: com.fasterxml.jackson.core.JsonParseException: Unrecognized token 'Foo': was expecting (JSON String, Number, Array, Object or token 'null', 'true' or 'false')
 at [Source: (String)"Foo{a=hello}"; line: 1, column: 4]
	at org.jooq.impl.JSONReader.read(JSONReader.java:113) ~[jooq-3.15.4.jar:na]
	at org.jooq.impl.DefaultBinding$DefaultResultBinding.readMultiset(DefaultBinding.java:3829) ~[jooq-3.15.4.jar:na]
	at org.jooq.impl.DefaultBinding$DefaultResultBinding.get0(DefaultBinding.java:3808) ~[jooq-3.15.4.jar:na]
	at org.jooq.impl.DefaultBinding$DefaultResultBinding.get0(DefaultBinding.java:3786) ~[jooq-3.15.4.jar:na]
	at org.jooq.impl.DefaultBinding$AbstractBinding.get(DefaultBinding.java:946) ~[jooq-3.15.4.jar:na]
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.setValue(CursorImpl.java:1551) ~[jooq-3.15.4.jar:na]
	... 96 common frames omitted
```

The problem boils down to the fact that the converter is called multiple times, similarly to #12413 or #12155: Once with the correct JSON string, but then again with the (toString()'ed) already converted `Foo` instance, which causes the JSON parser to fail.

Also possibly related: #12276 

### Steps to reproduce the problem

Run the given test method.

### Versions

- jOOQ: 3.15.4
- Java: openjdk version "11.0.11" 2021-04-20

**Comment by lukaseder** [comment]
Thanks a lot for your report. I can reproduce it, and indeed, this is due to a double conversion, which is wrong. Will look into fixing this right away.

**Comment by lukaseder** [comment]
Once fixed, there's a similar bug originating from debug logging the output:

```
PostgresTest.parseJson(JSON, Class<T>) line: 3913	
PostgresTest.lambda$31(JSON) line: 3925	
0x00000008010777c0.apply(Object) line: not available	
Converter$1.from(T) line: 178	
ConvertedDataType<T,U>.convert(Object) line: 255	
DSL.val(Object, DataType<T>) line: 29503	
DSL.val(Object, Field<T>) line: 29472	
Tools.field(Object, Field<T>) line: 1773	
RecordImpl1<T1>.valuesRow() line: 74	
RecordImpl1<T1>.valuesRow() line: 1	
AbstractResult<R>.format0(Object, boolean, boolean) line: 1364	
AbstractResult<R>.lambda$1(boolean, Object) line: 1360	
0x0000000801105808.apply(Object) line: not available	
ReferencePipeline$3$1.accept(P_OUT) line: 197	
ArrayList$Itr.forEachRemaining(Consumer<? super E>) line: 1003	
Spliterators$IteratorSpliterator<T>.forEachRemaining(Consumer<? super T>) line: 1845	
ReferencePipeline$3(AbstractPipeline<E_IN,E_OUT,S>).copyInto(Sink<P_IN>, Spliterator<P_IN>) line: 509	
ReferencePipeline$3(AbstractPipeline<E_IN,E_OUT,S>).wrapAndCopyInto(S, Spliterator<P_IN>) line: 499	
ReduceOps$3(ReduceOps$ReduceOp<T,R,S>).evaluateSequential(PipelineHelper<T>, Spliterator<P_IN>) line: 921	
ReferencePipeline$3(AbstractPipeline<E_IN,E_OUT,S>).evaluate(TerminalOp<E_OUT,R>) line: 234	
ReferencePipeline$3(ReferencePipeline<P_IN,P_OUT>).collect(Collector<? super P_OUT,A,R>) line: 682	
AbstractResult<R>.format0(Object, boolean, boolean) line: 1360	
ResultImpl<R>(AbstractResult<R>).format(Writer, TXTFormat) line: 197	
ResultImpl<R>(AbstractFormattable).format(TXTFormat) line: 117	
LoggerListener.log(Configuration, Result<?>) line: 188	
LoggerListener.resultEnd(ExecuteContext) line: 171	
ExecuteListeners.resultEnd(ExecuteContext) line: 252	
CursorImpl<R>.fetchNext(int) line: 240	
CursorImpl<R>(AbstractCursor<R>).fetch(int) line: 177	
CursorImpl<R>(AbstractCursor<R>).fetch() line: 88	
SelectQueryImpl<R>(AbstractResultQuery<R>).execute(ExecuteContext, ExecuteListener) line: 259	
SelectQueryImpl<R>(AbstractQuery<R>).execute() line: 335	
SelectQueryImpl<R>(AbstractResultQuery<R>).fetch() line: 284	
SelectImpl<R,T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,T14,T15,T16,T17,T18,T19,T20,T21,T22>.fetch() line: 2843	
PostgresTest.multipleConversionsInMultisetWithMappedJson() line: 3929	
NativeMethodAccessorImpl.invoke0(Method, Object, Object[]) line: not available [native method]	
NativeMethodAccessorImpl.invoke(Object, Object[]) line: 77	
DelegatingMethodAccessorImpl.invoke(Object, Object[]) line: 43	
Method.invoke(Object, Object...) line: 568	
FrameworkMethod$1.runReflectiveCall() line: 59	
FrameworkMethod$1(ReflectiveCallable).run() line: 12	
FrameworkMethod.invokeExplosively(Object, Object...) line: 56	
InvokeMethod.evaluate() line: 17	
RunBefores.evaluate() line: 26	
RunAfters.evaluate() line: 27	
TestWatcher$1.evaluate() line: 61	
ParentRunner$3.evaluate() line: 306	
BlockJUnit4ClassRunner$1.evaluate() line: 100	
JUnit4(ParentRunner<T>).runLeaf(Statement, Description, RunNotifier) line: 366	
JUnit4(BlockJUnit4ClassRunner).runChild(FrameworkMethod, RunNotifier) line: 103	
JUnit4(BlockJUnit4ClassRunner).runChild(Object, RunNotifier) line: 63	
ParentRunner$4.run() line: 331	
ParentRunner$1.schedule(Runnable) line: 79	
JUnit4(ParentRunner<T>).runChildren(RunNotifier) line: 329	
ParentRunner<T>.access$100(ParentRunner, RunNotifier) line: 66	
ParentRunner$2.evaluate() line: 293	
RunBefores.evaluate() line: 26	
RunAfters.evaluate() line: 27	
ParentRunner$3.evaluate() line: 306	
JUnit4(ParentRunner<T>).run(RunNotifier) line: 413	
JUnit4TestReference.run(TestExecution) line: 93	
TestExecution.run(ITestReference[]) line: 40	
RemoteTestRunner.runTests(String[], String, TestExecution) line: 529	
RemoteTestRunner.runTests(TestExecution) line: 756	
RemoteTestRunner.run() line: 452	
RemoteTestRunner.main(String[]) line: 210	
```	

I've created a separate issue for that: https://github.com/jOOQ/jOOQ/issues/12713

**Comment by lukaseder** [comment]
Note that the underlying issue of many of these is discussed also here: https://github.com/jOOQ/jOOQ/issues/11823

`convertFrom(Function<T, U>)` can be problematic in some edge cases, which are all bugs, of course, but if in doubt, the workaround can be to supply the `<U>` data type explicitly via `convertFrom(Class<U>, Funciton<T, U>)`. In your case, as a workround, try this:

```java
@Test
public void multipleConversionsInMultisetWithMappedJson() {
  create.select(
    multiset(
      select(
        // Pass Foo.class here to convertFrom()
        jsonObject(key("a").value("hello")).convertFrom(Foo.class, json -> parseJson(json, Foo.class))
      )
    )
  )
  .fetch();
}
```

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.16.0 and 3.15.6 (https://github.com/jOOQ/jOOQ/issues/12716). Thanks again for your report!
